### PR TITLE
Add hcloud support

### PR DESCRIPTION
**Pull request by bastelfreak**
This adds support for the Hetzner Cloud platform. They currently support
(2022-12-07) the folling images:

```
bastelfreak@bastelfreak-nb ~ $ hcloud image list --type system
ID         TYPE     NAME              DESCRIPTION       IMAGE SIZE   DISK SIZE   CREATED                         DEPRECATED
3          system   centos-7          CentOS 7          -            5 GB        Mon Jan 15 12:34:45 CET 2018    -
168855     system   ubuntu-18.04      Ubuntu 18.04      -            5 GB        Wed May  2 13:02:30 CEST 2018   -
5924233    system   debian-10         Debian 10         -            5 GB        Mon Jul  8 08:35:48 CEST 2019   -
15512617   system   ubuntu-20.04      Ubuntu 20.04      -            5 GB        Thu Apr 23 19:55:14 CEST 2020   -
45557056   system   debian-11         Debian 11         -            5 GB        Mon Aug 16 13:12:01 CEST 2021   -
45778012   system   centos-stream-8   CentOS Stream 8   -            5 GB        Thu Aug  5 07:07:23 CEST 2021   -
45780948   system   rocky-8           Rocky Linux 8     -            5 GB        Thu Aug 19 08:30:23 CEST 2021   -
59752342   system   centos-stream-9   CentOS Stream 9   -            5 GB        Thu Jan 27 08:52:03 CET 2022    -
67794396   system   ubuntu-22.04      Ubuntu 22.04      -            5 GB        Thu Apr 21 15:32:38 CEST 2022   -
69726282   system   fedora-36         Fedora 36         -            5 GB        Wed May 11 07:50:00 CEST 2022   -
76766499   system   rocky-9           Rocky Linux 9     -            5 GB        Wed Jul 20 15:55:52 CEST 2022   -
bastelfreak@bastelfreak-nb ~ $
```

### CODE REVIEW

**Comment by bastelfreak** [comment]
```
BEAKER_PUPPET_COLLECTION=puppet6 BEAKER_setfile=ubuntu1804-64{hostname=ubuntu1804-64.example.com} BEAKER_HYPERVISOR=hcloud bundle exec rake beaker
```

brings me:

```
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
/usr/bin/ruby -I/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib:/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-support-3.12.0/lib /home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/exe/rspec --pattern spec/acceptance/\*\*/\*_spec.rb
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-rspec-7.1.0/lib/beaker-rspec/helpers/serverspec.rb:43: warning: already initialized constant Module::VALID_OPTIONS_KEYS
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/specinfra-2.83.4/lib/specinfra/configuration.rb:4: warning: previous definition of VALID_OPTIONS_KEYS was here

Hosts file 'ubuntu1804-64{hostname=ubuntu1804-64.example.com}' does not exist.
Trying as beaker-hostgenerator input.

Hypervisor for ubuntu1804-64.example.com is hcloud
Beaker::Hypervisor, found some hcloud boxes to create
Provisioning hcloud
Generating SSH keypair
Creating hcloud SSH key
provisioning ubuntu1804-64.example.com
invalid input in fields 'source_server', 'image'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/hcloud-1.1.0/lib/hcloud/typhoeus_ext.rb:54:in `check_for_error'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/hcloud-1.1.0/lib/hcloud/client.rb:150:in `block in request'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/callbacks.rb:146:in `block in execute_callbacks'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/callbacks.rb:145:in `each'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/callbacks.rb:145:in `execute_callbacks'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/operations.rb:35:in `finish'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/easy_factory.rb:164:in `block in set_callback'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/ethon-0.16.0/lib/ethon/easy/response_callbacks.rb:74:in `block in complete'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/ethon-0.16.0/lib/ethon/easy/response_callbacks.rb:74:in `each'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/ethon-0.16.0/lib/ethon/easy/response_callbacks.rb:74:in `complete'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/ethon-0.16.0/lib/ethon/easy/operations.rb:34:in `perform'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/operations.rb:16:in `run'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/cacheable.rb:18:in `run'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/block_connection.rb:31:in `run'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/stubbable.rb:25:in `run'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/typhoeus-1.4.0/lib/typhoeus/request/before.rb:26:in `run'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/hcloud-1.1.0/lib/hcloud/client.rb:117:in `prepare_request'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/hcloud-1.1.0/lib/hcloud/abstract_resource.rb:9:in `prepare_request'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/hcloud-1.1.0/lib/hcloud/server_resource.rb:18:in `create'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/bundler/gems/beaker-hcloud-6d0512902e85/lib/beaker/hypervisor/hcloud.rb:76:in `create_server'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/bundler/gems/beaker-hcloud-6d0512902e85/lib/beaker/hypervisor/hcloud.rb:28:in `block in provision'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/bundler/gems/beaker-hcloud-6d0512902e85/lib/beaker/hypervisor/hcloud.rb:27:in `each'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/bundler/gems/beaker-hcloud-6d0512902e85/lib/beaker/hypervisor/hcloud.rb:27:in `provision'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-4.38.1/lib/beaker/hypervisor.rb:37:in `create'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-4.38.1/lib/beaker/network_manager.rb:72:in `block in provision'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-4.38.1/lib/beaker/network_manager.rb:71:in `each_key'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-4.38.1/lib/beaker/network_manager.rb:71:in `provision'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-rspec-7.1.0/lib/beaker-rspec/beaker_shim.rb:35:in `provision'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-rspec-7.1.0/lib/beaker-rspec/spec_helper.rb:57:in `block in <top (required)>'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core.rb:98:in `configure'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-rspec-7.1.0/lib/beaker-rspec/spec_helper.rb:5:in `<top (required)>'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-rspec-7.1.0/lib/beaker-rspec.rb:5:in `require'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-rspec-7.1.0/lib/beaker-rspec.rb:5:in `<module:BeakerRSpec>'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/beaker-rspec-7.1.0/lib/beaker-rspec.rb:1:in `<top (required)>'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/voxpupuli-acceptance-1.2.0/lib/voxpupuli/acceptance/spec_helper_acceptance.rb:12:in `require'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/voxpupuli-acceptance-1.2.0/lib/voxpupuli/acceptance/spec_helper_acceptance.rb:12:in `configure_beaker'
/home/bastelfreak/code/puppet-hdm/spec/spec_helper_acceptance.rb:8:in `<top (required)>'
/home/bastelfreak/code/puppet-hdm/spec/acceptance/docker_spec.rb:3:in `require'
/home/bastelfreak/code/puppet-hdm/spec/acceptance/docker_spec.rb:3:in `<top (required)>'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/configuration.rb:2117:in `load'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/configuration.rb:2117:in `load_file_handling_errors'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/configuration.rb:1617:in `block in load_spec_files'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/configuration.rb:1615:in `each'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/configuration.rb:1615:in `load_spec_files'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/runner.rb:102:in `setup'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/runner.rb:86:in `run'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/runner.rb:71:in `run'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib/rspec/core/runner.rb:45:in `invoke'
/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/exe/rspec:4:in `<main>'
/usr/bin/ruby -I/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/lib:/home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-support-3.12.0/lib /home/bastelfreak/code/puppet-hdm/.vendor/ruby/3.0.0/gems/rspec-core-3.12.0/exe/rspec --pattern spec/acceptance/\*\*/\*_spec.rb failed
```

**Review by ekohl — COMMENTED** [review]
We also do some transformations specific to a hypervisor:
https://github.com/voxpupuli/beaker-hostgenerator/blob/master/lib/beaker-hostgenerator/hypervisor/docker.rb for example. It looks to be very consistent `NAME-VERSION` without an architecture so perhaps that's an easier solution.

**Comment by bastelfreak** [comment]
aha!
```
$ bundle exec beaker-hostgenerator --hypervisor hcloud ubuntu2004-64
---
HOSTS:
  ubuntu2004-64-1:
    platform: ubuntu-20.04-amd64
    image: ubuntu-20.04
    hypervisor: hcloud
    roles:
    - agent
CONFIG:
  nfs_server: none
  consoleport: 443
```

**Comment by ekohl** [comment]
> a problem I noticed: We generate an FQDN like this: ubuntu1804-64.example.com. We will create two instances with that name, one for Puppet 6 and one for Puppet 7. At Hetzner, a project cannot contain multiple virtual machines with the same FQDN. If we assume we've one project for the whole Vox Pupuli organisation, the name needs to contain puppet+PR info, or a random string? I don't think that the name is actually important for the running tests.

FQDN generation is something we now do here: https://github.com/voxpupuli/puppet_metadata/blob/1d92f203598362d9719971cb5b6cab6c085f9689/lib/puppet_metadata/beaker.rb#L48

I'd actually prefer some option to set this and an env var like `BEAKER_HOSTGENERATOR_DOMAIN` where beaker-hostgenerator generates the FQDN.

Then for the hcloud hypervisor it could actually add some randomness to it. For reference, libvirt has this: https://github.com/voxpupuli/beaker-vagrant/blob/50d671f73862e7a2271b1b361036794489a64672/lib/beaker/hypervisor/vagrant_libvirt.rb#L7-L11

**Comment by bastelfreak** [comment]
@ekohl I considered modifying the image name in the hcloud hypervisor file. but hcloud only supports a tiny subset of all the operating systems beaker-hostgenerator has. Filtering that might add a bit too much complexity. Or do you think we shouldn't filter that and generate all operating systems for hcloud?

**Comment by ekohl** [comment]
> Or do you think we shouldn't filter that and generate all operating systems for hcloud?

I think this. Just like we generate a docker image name on the assumption that it exists. It may exist in the future and otherwise you get an error from the hypervisor. For example, it's reasonable to assume that at some point there will be an image for Ubuntu 24.04 and it's nice not to need any code change for it when it comes out.

**Comment by bastelfreak** [comment]
@ekohl I refactored the image generation. I just need to think about the FQDN now.

File: lib/beaker-hostgenerator/hypervisor/hcloud.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    class Hcloud < BeakerHostGenerator::Hypervisor::Interface
+      include BeakerHostGenerator::Data
+
+      def generate_node(node_info, base_config, bhg_version)
+        # split ostype into string and array
+        # ostype is ubuntu2204, debian10, rocky9
+        os, version = node_info['ostype'].split(/(\D+)/).reject!(&:empty?)
+        base_config['image'] = case os
+        when 'ubuntu'
+          puts version
```
Comment (ekohl) [inline lib/beaker-hostgenerator/hypervisor/hcloud.rb:17]:
```suggestion
```

File: lib/beaker-hostgenerator/hypervisor/hcloud.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+          "#{os}-#{version[0, 2]}.#{version[2,2]}"
+        when 'centos'
+          [8, 9].include?(os) ? "#{os}-stream-#{version}" : "#{os}-#{version}"
```
Comment (ekohl) [inline lib/beaker-hostgenerator/hypervisor/hcloud.rb:20]: I'd go the other way around: there's just CentOS 7 (all older versions are EOL) which isn't stream and 10 will likely also include stream.

File: lib/beaker-hostgenerator/hypervisor.rb [hunk 3]
```diff
@@ -45,7 +45,8 @@ def self.builtin_hypervisors()
         'vagrant' => BeakerHostGenerator::Hypervisor::Vagrant,
         'vagrant_libvirt' => BeakerHostGenerator::Hypervisor::Vagrant,
         'docker' => BeakerHostGenerator::Hypervisor::Docker,
-        'abs' => BeakerHostGenerator::Hypervisor::ABS
+        'abs' => BeakerHostGenerator::Hypervisor::ABS,
+        'hcloud' => BeakerHostGenerator::Hypervisor::Hcloud
```
Comment (ekohl) [inline lib/beaker-hostgenerator/hypervisor.rb:49]: Does Rubocop allow a trailing comma?
```suggestion
        'hcloud' => BeakerHostGenerator::Hypervisor::Hcloud,
```

File: lib/beaker-hostgenerator/hypervisor/hcloud.rb (hunk 2 above)
Comment (bastelfreak) [reply lib/beaker-hostgenerator/hypervisor/hcloud.rb:20]: ah okay. I assumed `stream` will be dropped for 10. In that case I will revert the logic.

File: lib/beaker-hostgenerator/hypervisor/hcloud.rb (hunk 1 above)
Comment (bastelfreak) [reply lib/beaker-hostgenerator/hypervisor/hcloud.rb:17]: ups, that's from my debugging.

File: lib/beaker-hostgenerator/hypervisor/hcloud.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+      def generate_node(node_info, base_config, bhg_version)
+        # split ostype into string and array
+        # ostype is ubuntu2204, debian10, rocky9
+        os, version = node_info['ostype'].split(/(\D+)/).reject!(&:empty?)
+        base_config['image'] = case os
+        when 'ubuntu'
+          "#{os}-#{version[0, 2]}.#{version[2,2]}"
+        when 'centos'
+          version.to_i == 7 ? "#{os}-#{version}" : "#{os}-stream-#{version}"
```
Comment (bastelfreak) [inline lib/beaker-hostgenerator/hypervisor/hcloud.rb:19]: @ekohl I reversed the logic and also added a test case for centos7/8 to verify it's working correctly.

File: lib/beaker-hostgenerator/hypervisor/hcloud.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+require 'beaker-hostgenerator/data'
+require 'beaker-hostgenerator/hypervisor'
+require 'deep_merge/rails_compat'
```
Comment (ekohl) [inline lib/beaker-hostgenerator/hypervisor/hcloud.rb:3]: Why is this needed? I don't see any merging going on.

Comment (bastelfreak) [reply lib/beaker-hostgenerator/hypervisor/hcloud.rb]: probably a leftover because I started with the docker implementation as base for hcloud. I removed it.
